### PR TITLE
Aims input sets

### PR DESCRIPTION
**Pull request by tpurcell90**
## Summary

Adds the `aims_input_set` from atomate2 into pymatgen

Major changes:

- feature 1: AimsInputSet is now here

## Checklist

- [x] Google format doc strings added. Check with `ruff`.
- [x] Type annotations included. Check with `mypy`.
- [x] Tests added for new features/fixes.
- [x] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

### CODE REVIEW

**Comment by tpurcell90** [comment]
Is there something I have to do in my personal fork to prevent the VASP errors when running the CI tests?

**Comment by janosh** [comment]
@tpurcell90 This error is genuinely yours:

```py
        properties = ("energy", "free_energy", "forces")
    
        in_set = AimsInputSet(parameters, Si, properties)
        assert check_file(geometry_in_str, in_set.geometry_in.get_str())
        assert check_file(control_in_str, in_set.control_in.get_str())
>       assert parameters_json_str == in_set.parameters_json
E       assert '{\n  "xc": "...    2\n  ]\n}' == '{\n  "xc": "...    2\n  ]\n}'
E           {
E             "xc": "pbe",
E         -   "species_dir": "D:\\a\\pymatgen\\pymatgen\\tests\\io\\aims\\species_directory\\light",
E         ?                      - -          -         -      -  -     -                   -
E         +   "species_dir": "D:\a\pymatgen\pymatgen\tests\io\aims\species_directory\light",
E             "k_grid": [
E               2,
E               2,
E               2
E             ]
E           }

tests\io\aims\test_input_set.py:278: AssertionError
```

The others are also on the `master` branch and something I need to find time to fix. Nothing to worry about.

**Comment by tpurcell90** [comment]
I moved in all of the input generator stuff from atomate2 into here.

Should I make a separate conftest file for aims?

**Comment by tpurcell90** [comment]
I just saw the windows error If this does not pass I will try to compile everything on windows and see what I need to do to solve it.

**Comment by tpurcell90** [comment]
everything should now be ready for review

**Comment by tpurcell90** [comment]
Ready to merge?

**Comment by tpurcell90** [comment]
Do I need to make any changes to the code?

**Comment by janosh** [comment]
@tpurcell90 Thanks for this contribution and sorry for the slow reply. I haven't had time to review this lately. I hope I'll get to it this week. Feel free to ping.

File: tests/io/aims/test_sets/test_bs_generator.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    check.pop("species_dir", None)
+    check_output = check.pop("output", None)
+
+    assert ref == check
+
+    if check_output:
+        for ref_out, check_out in zip(ref_output, check_output):
+            if "band" in check_out:
+                check_band(check_out, ref_out)
```
Comment (janosh) [inline tests/io/aims/test_sets/test_bs_generator.py:52]: this function doesn't raise and its return value is not used so this test likely isn't working as intended

File: tests/io/aims/test_sets/test_static_restart_from_relax_generator.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        "static-from-prev-default-sd-o2",
+        tmp_path,
+        ref_path,
+        species_dir,
+    )
+
+    if sd_def:
+        os.environ["AIMS_SPECIES_DIR"] = sd_def
+    else:
+        os.unsetenv("AIMS_SPECIES_DIR")
```
Comment (janosh) [inline tests/io/aims/test_sets/test_static_restart_from_relax_generator.py:134]: you can use the `pytest` `monkeypatch` fixture to temp override dict keys or class attributes

**Review by janosh — CHANGES_REQUESTED** [review]
do we actually need this many new test files? are they all needed to test separate functionality?

**Review by janosh — CHANGES_REQUESTED** [review]
do we actually need this many new test files? are they all needed to test separate functionality?

left a few comments in the code

**Comment by tpurcell90** [comment]
> do we actually need this many new test files? are they all needed to test separate functionality?
> 
> left a few comments in the code

The tests are for all possible InputSet combinations to make sure none of the corner cases break

File: tests/io/aims/test_sets/test_static_restart_from_relax_generator.py (hunk 2 above)
Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_restart_from_relax_generator.py]: Right I was not thinking of that when doing this test (first developed it outside of pytest)

File: tests/io/aims/test_sets/test_static_generator.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    with open(f"{ref_dir / test_name}/parameters.json") as ref_file:
+        ref = json.load(ref_file)
+    ref.pop("species_dir", None)
+
+    with open(f"{work_dir / test_name}/parameters.json") as check_file:
+        check = json.load(check_file)
+    check.pop("species_dir", None)
+
+    assert ref == check
```
Comment (janosh) [inline tests/io/aims/test_sets/test_static_generator.py:30]: there's some redundancy between the test files in this directory. they all define nearly or even fully identical `compare_files` functions. can we put this function into an `__init__.py` file and import `compare_files` and other helpers from there?

File: pymatgen/io/aims/sets/base.py [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+
+    @classmethod
+    def from_str(cls, contents: str):
+        """Create an input file from the contents string.
+
+        Args:
+            contents (str): The contents of the input file
+        """
+        return cls(contents)
+
```
Comment (janosh) [inline pymatgen/io/aims/sets/base.py:66]: this class looks a bit like boilerplate for a simple `str`. can we just replace it with `content_str` itself?

Comment (tpurcell90) [reply pymatgen/io/aims/sets/base.py:66]: It is boilerplate, but that is how the class is defined in `io.core.InputFile`  so I used the same template. I can change it, but I would either have to remove the `InputFile` inheritance since those methods are abstract in `InputFile` or modify `InputFile` itself.
 https://github.com/materialsproject/pymatgen/blob/0d7e6cd3b7800664e850938a7ebfe1240a3bd319/pymatgen/io/core.py#L48

File: tests/io/aims/test_sets/test_static_generator.py (hunk 3 above)
Comment (janosh) [reply tests/io/aims/test_sets/test_static_generator.py:30]: we'll figure it out. happy to help if errors come up :)

File: pymatgen/io/aims/sets/base.py (hunk 4 above)
Comment (janosh) [reply pymatgen/io/aims/sets/base.py:66]: i wouldn't use a class at all here. note the `InputFile` mentions its use is optional

> Abstract base class to represent a single input file. Note that use of this class is optional; it is possible create an InputSet that does not rely on underlying InputFile objects.

File: tests/io/aims/test_sets/test_static_generator.py (hunk 3 above)
Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_generator.py:30]: Now I remember what the error is, one of the packages installed using `pip install .[strict]` creates a `tests` directory in `site-packages` which conflicts with the `tests/__init__.py` file. I am in the process of seeing if I can figure out which package it is.

Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_generator.py:30]: The offending package is `pybtex==0.24.0`

Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_generator.py:30]: I'm not sure how to override this without some sort of very hacky workaround with the sys.path

File: pymatgen/io/aims/sets/base.py (hunk 4 above)
Comment (tpurcell90) [reply pymatgen/io/aims/sets/base.py:66]: I can remove it then, I included it to match the other codes io scheme

File: tests/io/aims/test_sets/test_static_generator.py (hunk 3 above)
Comment (janosh) [reply tests/io/aims/test_sets/test_static_generator.py:30]: you could try [this SO answer](https://stackoverflow.com/a/46976704), or turn the helper function into a `pytest` fixture in wich case there's no need to import it. it'll be available as a function argument to each test function.

File: pymatgen/io/aims/sets/base.py (hunk 4 above)
Comment (janosh) [reply pymatgen/io/aims/sets/base.py:66]: haven't looked very closely but wouldn't be surprised if some of the other codes could do without `InputFile` as well

File: tests/io/aims/test_sets/test_static_generator.py (hunk 3 above)
Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_generator.py:30]: Okay I will try the pytest fixture route first

File: pymatgen/io/aims/sets/base.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+            if aims_name is not None:
+                updated_params[aims_name] = True
+
+        aims_geometry_in = AimsGeometryIn.from_structure(self._structure)
+        aims_control_in = AimsControlIn(updated_params)
+
+        os.makedirs(TMPDIR_NAME, exist_ok=True)
+        with monty.os.cd(TMPDIR_NAME):
+            aims_control_in.write_file(self._structure)
+            aims_control_in_content = AimsInputFile.from_file("control.in")
```
Comment (tpurcell90) [inline pymatgen/io/aims/sets/base.py:121]: @janosh This might be the reason for using `InputFiles` instead of just strings so there is a way to automatically read in a file as is. I can still remove it but the `from_file` calls from `InputFile` are convenient

Comment (janosh) [reply pymatgen/io/aims/sets/base.py:121]: all that does is call `InputFile.from_file` which are just these 2 lines

```py
with zopen(filename, mode="rt") as f:
    return cls.from_str(f.read())
```

IMO doesn't deserve a whole wrapper class. might as well do

```py
with zopen(filename, mode="rt") as f:
    aims_control_in_content = f.read()
```

Comment (tpurcell90) [reply pymatgen/io/aims/sets/base.py:121]: True, for FHI-aims it is  not an issue. I will make the change today.

**Comment by tpurcell90** [comment]
Ready to merge?

**Comment by janosh** [comment]
while i really appreciate the thorough test coverage, i'm a bit reluctant to merge so many new test files. most of them are small which makes them unproblematic but several like `tests/io/aims/aims_input_generator_ref/relax-default-sd-si/aims.out.gz` are ~50KB. still not a huge deal but you sure there's no redundancy among them?

**Comment by tpurcell90** [comment]
> while i really appreciate the thorough test coverage, i'm a bit reluctant to merge so many new test files. most of them are small which makes them unproblematic but several like `tests/io/aims/aims_input_generator_ref/relax-default-sd-si/aims.out.gz` are ~50KB. still not a huge deal but you sure there's no redundancy among them?

I have condensed down the test files as much as possible without harming the coverage.

**Review by janosh — CHANGES_REQUESTED** [review]
looking much better now ito files added. let me know if you're ok with the changes in 
e44f52d

File: tests/io/aims/test_sets/test_static_restart_from_relax_generator.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        f"{ref_path}/relax-no-kgrid-si/",
+        "static-from-prev-no-kgrid-si",
+        tmp_path,
+        ref_path,
+        species_dir,
+    )
+
+
+def test_static_from_relax_default_species_dir(tmp_path, monkeypatch):
+    monkeypatch.setenv("AIMS_SPECIES_DIR", str(species_dir / "light"))
```
[the remark this answers is not in the document]
Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_restart_from_relax_generator.py:57]: Those changes look good, and I can make that change

[the remark this answers is not in the document]
Comment (tpurcell90) [reply tests/io/aims/test_sets/test_static_restart_from_relax_generator.py]: I made the conftest changes

File: pymatgen/util/testing/aims.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    if check_output:
+        for ref_out, check_out in zip(ref_output, check_output):
+            if "band" in check_out:
+                assert check_band(check_out, ref_out)
+            else:
+                assert ref_out == check_out
+
+
+def comp_system(
```
Comment (janosh) [inline pymatgen/util/testing/aims.py:64]: you're defining two `comp_system` functions. i think the one in `tests/io/aims/test_sets/test_static_restart_from_relax_generator.py#L12` can go and be replaced by this one?

Comment (janosh) [reply pymatgen/util/testing/aims.py:64]: also, feel free to improve the doc strings on these helper methods. i added very generic ones but you have a better understanding of what they do

Comment (tpurcell90) [reply pymatgen/util/testing/aims.py:64]: The two are slightly different, but I can combine them

I'll add more complete docstrings now that I am in the office

Comment (tpurcell90) [reply pymatgen/util/testing/aims.py:78]: I made these two fixes

**Review by janosh — APPROVED** [review]
thanks so much for your patience and for addressing all my comments in record time! this a major contribution and hugely appreciated! `pymatgen` would not be the tool it is today without all the effort put in by volunteers like you!

btw, i just noticed you're not [listed as a contributor](https://pymatgen.org/team) yet. feel free to submit a PR to add yourself to `docs/assets/metadata.yml`

**Comment by JaGeo** [comment]
Huge thank you from a frequent pymatgen user for this large amount of work, @tpurcell90 !
